File: src/spoj/dynamic_programming/mnered/MNERED.c
// http://www.spoj.com/problems/MNERED/

#include <stdio.h>

#define DIMENSION 105

int field[DIMENSION][DIMENSION];
int dimension;
int amount_of_entities;

void calculate_cumulative_2d_matrix() {
	int y;
	int x;

	for(x = 1; x <= dimension; ++x) {
		for(y = 0; y <= dimension; ++y) {
			field[y][x] += field[y][x - 1];
		}
	}

	for(y = 1; y <= dimension; ++y) {
		for(x = 0; x <= dimension; ++x) {
			field[y][x] += field[y - 1][x];
		}
	}
}

int solve() {
	// w - width of sliding window
	// h - height of sliding window

	int w = 1;
	int h;
	int tmp;
	int y;
	int x;
	int curr_non_empty_cells_cnt;
	int max_non_empty_cells_cnt;

	while(w * w <= amount_of_entities + 1) {
		if(amount_of_entities % w == 0) {
			h = amount_of_entities / w;

			if((w <= dimension) && (h <= dimension)) {
				
				// horizontal sliding window
				for(y = h; y <= dimension; ++y) {
					for(x = w; x <= dimension; ++x) {
						curr_non_empty_cells_cnt =   field[y][x] 
												   - field[y - h][x] 
												   - field[y][x - w] 
												   + field[y - h][x - w];
					
						if(curr_non_empty_cells_cnt > max_non_empty_cells_cnt) {
							max_non_empty_cells_cnt = curr_non_empty_cells_cnt;
						}
					}
				}
			
				// swap w and h
				tmp = w;
				w = h;
				h = tmp;

				// vertical sliding window
				for(y = h; y <= dimension; ++y) {
					for(x = w; x <= dimension; ++x) {
						curr_non_empty_cells_cnt =   field[y][x] 
												   - field[y - h][x] 
												   - field[y][x - w] 
												   + field[y - h][x - w];
					
						if(curr_non_empty_cells_cnt > max_non_empty_cells_cnt) {
							max_non_empty_cells_cnt = curr_non_empty_cells_cnt;
						}
					}
				}

				// restore original value of w
				w = h;
			}
		}
		w++;
	}

	return amount_of_entities - max_non_empty_cells_cnt;
}

int main() {
	scanf("%d %d", &dimension, &amount_of_entities);

	int i;
	int x;
	int y;
	for(i = 0; i < amount_of_entities; ++i) {
		scanf("%d %d", &y, &x);
		field[y][x] = 1;
	}

	calculate_cumulative_2d_matrix();

	int result = solve();
	printf("%d\n", result);

	return 0;
}

File: src/spoj/dynamic_programming/mnered/MNERED.java
package spoj.dynamic_programming.mnered;

// http://www.spoj.com/problems/MNERED/

import java.util.Arrays;
import java.util.List;

public class MNERED {

	public static void main(String[] args) {

		// example from description of the problem
		// expected result is 2
		solve(4, Arrays.asList(
				new Coordinate(2, 2),
				new Coordinate(4, 4),
				new Coordinate(1, 1)));

		// example from description of the problem
		// expected result is 3
		solve(5, Arrays.asList(
				new Coordinate(2, 2),
				new Coordinate(3, 2),
				new Coordinate(4, 2),
				new Coordinate(2, 4),
				new Coordinate(3, 4),
				new Coordinate(4, 4),
				new Coordinate(2, 3),
				new Coordinate(2, 3)));

		// 9 entities
		// 2 entities - near each other
		// 7 entities on the top of one entity
		// expected result is 7
		solve(10, Arrays.asList(
				new Coordinate(2, 2),
				new Coordinate(2, 2),
				new Coordinate(2, 2),
				new Coordinate(2, 2),
				new Coordinate(2, 2),
				new Coordinate(2, 2),
				new Coordinate(2, 2),
				new Coordinate(2, 2),
				new Coordinate(3, 3)));

		// entities are already aligned into rectangle
		// expected result is 0
		solve(10, Arrays.asList(
				new Coordinate(2, 2),
				new Coordinate(2, 3),
				new Coordinate(2, 4),
				new Coordinate(3, 2),
				new Coordinate(3, 3),
				new Coordinate(3, 4),
				new Coordinate(4, 2),
				new Coordinate(4, 3),
				new Coordinate(4, 4)));
	}

	static void solve(int dimension, List<Coordinate> entitiesCoordinates) {

		// Put entities onto the field
		int[][] field = new int[dimension + 1][dimension + 1];
		for (Coordinate c : entitiesCoordinates) {
			field[c.y][c.x] = 1;
		}

		// Calculate 2D cumulative matrix
		for (int x = 1; x < field.length; x++) {
			for (int y = 0; y < field.length; y++) {
				field[y][x] += field[y][x - 1];
			}
		}
		for (int y = 1; y < field.length; y++) {
			for (int x = 0; x < field.length; x++) {
				field[y][x] += field[y - 1][x];
			}
		}

		// Check all rectangles
		int maxAmountOfNonEmptyCells = -1;
		int amountOfEntitites = entitiesCoordinates.size();
		int w = 1;
		while ((w * w) <= (amountOfEntitites + 1)) {
			if ((amountOfEntitites % w) == 0) {
				int h = amountOfEntitites / w;

				if ((w <= dimension) && (h <= dimension)) {

					// Horizontal sliding window
					for (int y = h; y < field.length; y++) {
						for (int x = w; x < field.length; x++) {
							int nonEmptyCells = (field[y][x]
									- field[y - h][x]
									- field[y][x - w])
									+ field[y - h][x - w];
							maxAmountOfNonEmptyCells = Math.max(maxAmountOfNonEmptyCells, nonEmptyCells);
						}
					}

					// Temporarily swap H and W
					int tmp = w;
					w = h;
					h = tmp;

					// Vertical sliding window
					for (int y = h; y < field.length; y++) {
						for (int x = w; x < field.length; x++) {
							int nonEmptyCells = (field[y][x]
									- field[y - h][x]
									- field[y][x - w])
									+ field[y - h][x - w];
							maxAmountOfNonEmptyCells = Math.max(maxAmountOfNonEmptyCells, nonEmptyCells);
						}
					}

					// Restore original values of H and W
					tmp = w;
					w = h;
					h = tmp;
				}
			}
			w++;
		}

		int result = amountOfEntitites - maxAmountOfNonEmptyCells;
		System.out.println(result);
	}

	/**
	 * x and y starts from 1 (not from 0)
	 */
	static class Coordinate {
		int x;
		int y;
		public Coordinate(int x, int y) {
			this.x = x;
			this.y = y;
		}
	}
}
